src/napari_imagej/_function.py

```diff
@@ -149,25 +178,41 @@ def _preprocess_remaining_inputs(
     """Resolves each input in unresolved_inputs"""
     resolved_java_args = _convert_inputs(user_resolved_inputs)
     # resolve remaining inputs
-    for i in range(len(unresolved_inputs)):
-        name = unresolved_inputs[i].getName()
-        obj = resolved_java_args[i]
-        if obj is None:
+    for module_item, input in zip(unresolved_inputs, resolved_java_args):
+        name = module_item.getName()
+        if input is None and module_item.isRequired():
             raise ValueError(
                 "No selection was made for input {}!".format(name)
             )
-        module.setInput(name, obj)
+        item_class = module_item.getType()
+        if not item_class.isInstance(input):
+            if ij.convert().supports(input, item_class):
+                input = ij.convert().convert(input, item_class)
+            else:
+                raise ValueError(
+                    f"{input} is not a {module_item.getType()}!"
+                )
+        module.setInput(name, input)
         module.resolveInput(name)
 
     # sanity check: ensure all inputs resolved
     for input in inputs:
         if input.isRequired() and not module.isInputResolved(input.getName()):
             raise ValueError(
-                "No selection was made for input {}!".format(input.getName())
+                f"input {input.getName()} of type {input.getType()} was not resolved! If it is impossible to resolve, let us know at forum.image.sc or by filing an issue at https://github.com/imagej/napari-imagej!".format(input.getName())
             )
 
     return resolved_java_args
 
+def _resolvable_or_required(input):
+    """Determines whether input should be resolved in napari"""
+    if input.isRequired(): return True
+    try:
```

Comment:
I don't like this `try`/`except` statement, we could be catching things that aren't due to a conversion error.

We could return a tuple `(type, ok)` like they often do in Go.

Reply:
Yeah, @gselzer, it's yucky, but there isn't much we can do about it.

This is the most correct way.

Reply:
Without fully understanding the design here yet, I feel like we could do better. I'll try to take a look while I am getting napari-imagej set up on my laptop tomorrow.

Reply:
@ctrueden what do you mean by "do better"? Not use the try/except?

I'd think so too, but we'd probably have to rewrite the `python_type_of` function, or at least define a helper that wraps this try/except

I'm open to any better ways, if you can find one.